### PR TITLE
impl Default for AddressSpace to make future changes easier

### PR DESCRIPTION
**Pull request by folkertdev**
in #383 it came up that the current representation of an address space is not entirely correct. To make changes to that type easier, and because usually "Generic" is the address space that users want, implementing and using the default instance means that even if we change the representation of address spaces, this would not actually be noticeable by users.

## \<Breaking Changes\>

There specifically are no breaking changes in the PR

## Checklist

- [x] I have read the [Contributing Guide](https://github.com/TheDan64/inkwell/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

**Comment by folkertdev** [comment]
cc @hedgar2017

**Comment by hedgar2017** [comment]
@folkertdev I think we must remove this enum completely and use u32 instead.

**Comment by hedgar2017** [comment]
@folkertdev each target has its own understanding and treatment of address spaces. For example for our virtual machine it's:
0 - stack
1 - the 1st heap
2 - the 2nd heap
3 - arbitrary memory page
4 - the space of constants
and so on.

Some targets may use a large number of address spaces, so someone will eventually be blocked by this enum anyway.

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
Let's remove the enum?
It's already casted to u32 everywhere.

**Comment by folkertdev** [comment]
I understand that you want to change the representation. That is however breaking for pretty much every user of inkwell. so this PR would be a step towards making that process easier, because usually you want to use the default address space.

Anyway, I don't know exactly what the policy is on breaking changes, that is ultimately up to @TheDan64 to decide. This PR is safe to merge no matter what, since it is strictly additive.

**Comment by hedgar2017** [comment]
@folkertdev I understand, but it's not a good reason to stockpile such blocking and misleading technical debt.

Concerning this patch, adding this default may be reasonable for your particular target, but for a lot of other targets the 0-th address space is not more default than others.

**Comment by folkertdev** [comment]
well, https://llvm-swift.github.io/LLVMSwift/Structs/AddressSpace.html says

> An address space is an identifier for a target-specific range of address values. An address space is a fundamental part of the type of a pointer value and the type of operations that manipulate memory.
> 
> LLVM affords a default address space (numbered zero) and places a number of assumptions on pointer values within that address space:
> 
> - The pointer must have a fixed integral value
> - The null pointer has a bit-value of 0

so out of all the address spaces, I think `0` is objectively the most default. For common targets, it is also probably the only one you'll ever use

**Comment by hedgar2017** [comment]
@folkertdev well, you're right, I indeed found this in the official doc.
It's up to @TheDan64 of course.

**Review by hedgar2017 — APPROVED** [review]

**Comment by TheDan64** [comment]
Has anyone confirmed 5+ are valid values LLVM accepts for address spaces?

**Comment by TheDan64** [comment]
I'm ok with converting to a u32 if so

**Comment by folkertdev** [comment]
well, it turns out that the address space is really as 24-bit integer. in other words, this test passes 

```rust
#[test]
fn test_ptr_address_space() {
    let context = Context::create();

    let spaces = [0, 1, 2, 3, 4, 5, 6, 1 << 24 - 1];

    for index in spaces {
        let address_space = AddressSpace(index);

        let ptr = context.i32_type().ptr_type(address_space);
        assert_eq!(ptr.get_address_space(), address_space);
    }

    let address_space = AddressSpace(1 << 24);
    let ptr = context.i32_type().ptr_type(address_space);
    assert_ne!(ptr.get_address_space(), address_space);
}
```

Not sure what to do with that. We could pick `u16` internally, and assume that is good enough (really how many address spaces can you use in practice). Using `u32` is incorrect I think and using it would not align with Inkwell's goals.

**Comment by hedgar2017** [comment]
@folkertdev thanks for checking! By the way, we can use `2^24 - 1` address spaces in practice. If you don't use them, it doesn't mean that nobody else needs it.

I propose we make some kind of `AddressSpace(u32)` wrapper with an assertion in the constructor that it's not bigger than `2^24 - 1`. This way we'll fully cover the functionality LLVM provides and still keep it safe and ugly-low-level-crash-free.

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
Oh, thanks a lot for this change.
Please encapsulate the inner value with an `address_space < 2^24` assertion and I believe it's ready for merging!

File: src/lib.rs
Comment (hedgar2017) [inline src/lib.rs:125]: These constants are probably redundant. Do you need them?

**Review by hedgar2017 — CHANGES_REQUESTED** [review]

File: src/lib.rs
Comment (hedgar2017) [reply src/lib.rs:125]: Another alternative is not using magic numbers and having an enum in your own front-end with proper descriptive names for each address space.
`inkwell::AddressSpace::try_from(crate::AddressSpace::Stack)?` is not ugly.

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
I believe we should remove the constants, because their names are not descriptive like @folkertdev himself said a while ago.
The inkwell users must define their AddressSpace logic outside of this crate.

**Comment by folkertdev** [comment]
I disagree. you would force functions upon your users that look like they can panic, but they actually can't. That is disappointing API design.

**Comment by hedgar2017** [comment]
I didn't get what the constants had to do with panics.
No big deal for me, it's up to @TheDan64

**Comment by folkertdev** [comment]
added an instance for `From<u16>` (that means `::from(3)` works and is not ambiguous. `u8` can always (and conveniently) be upcast to `u16`, and removed the constants.

**Comment by hedgar2017** [comment]
@folkertdev has your commit passed through? I see the last one 21h ago.

**Comment by folkertdev** [comment]
should be good now, I think this is ready.

**Review by hedgar2017 — APPROVED** [review]

**Review by TheDan64 — APPROVED** [review]
Thanks!

**Pull request merged**
